internal/client/rpc_sender.go

```diff
@@ -24,47 +24,43 @@ import (
 
 	"github.com/cockroachdb/cockroach/roachpb"
 	"github.com/cockroachdb/cockroach/rpc"
+	"github.com/pkg/errors"
 )
 
 const healthyTimeout = 2 * time.Second
 
 type sender struct {
-	conn   *grpc.ClientConn
-	client roachpb.ExternalClient
+	remoteAddr string
+	conn       *grpc.ClientConn
+	client     roachpb.ExternalClient
+	rpcContext *rpc.Context
 }
 
 // NewSender returns an implementation of Sender which exposes the Key-Value
 // database provided by a Cockroach cluster by connecting via RPC to a
 // Cockroach node.
-func NewSender(ctx *rpc.Context, target string) (Sender, error) {
-	conn, err := ctx.GRPCDial(target)
+func NewSender(ctx *rpc.Context, remoteAddr string) (Sender, error) {
+	conn, err := ctx.GRPCDial(remoteAddr)
 	if err != nil {
 		return nil, err
 	}
 	return &sender{
-		conn:   conn,
-		client: roachpb.NewExternalClient(conn),
+		remoteAddr: remoteAddr,
+		conn:       conn,
+		client:     roachpb.NewExternalClient(conn),
+		rpcContext: ctx,
 	}, nil
 }
 
 // Send implements the Sender interface.
 func (s *sender) Send(ctx context.Context, ba roachpb.BatchRequest) (*roachpb.BatchResponse, *roachpb.Error) {
-	ctxWithTimeout, cancel := context.WithTimeout(ctx, healthyTimeout)
-	defer cancel()
-
-	c := s.conn
-	for state, err := c.State(); state != grpc.Ready; state, err = c.WaitForStateChange(ctxWithTimeout, state) {
-		if err != nil {
-			return nil, roachpb.NewErrorf("roachpb.Batch RPC failed: %s", err)
-		}
-		if state == grpc.Shutdown {
-			return nil, roachpb.NewErrorf("roachpb.Batch RPC failed as client connection was closed")
-		}
+	if err := s.rpcContext.WaitForConnected(s.remoteAddr, healthyTimeout); err != nil {
```

Comment:
shouldn't `WaitForConnected` take the `context.Context` and fail early when it expires?